accepted/future-releases/base-interface-final/feature-specification.md

```diff
@@ -0,0 +1,596 @@
+without *intending* it to be used as a mixin and then accidentally forbid that
+usage by adding a generative constructor or superclass to the class. That is a
+breaking change to any downstream user that had that class in a `with` clause.
+
+Eventually, we would like classes to not be usable as a mixin by default. Being
+extensible and implementable seems to be the right default for classes based on
+Dart's history and talking to users. But being usable as a mixin by default is
+rarely (but not *never*) useful and often confusing. It makes classes brittle
+for little upside.
+
+Instead, under this proposal we require authors to explicitly opt in to allowing
+the class to be used as a mixin by adding a `mixin` modifier to the class:
+
+```dart
+class OnlyClass {}
+
+class FailUseAsMixin extends OtherSuperclass with OnlyClass {} // Error.
+
+mixin class Both {}
+
+class UsesAsSuperclass extends Both {}
+
+class UsesAsMixin extends OtherSuperclass with Both {} // OK.
+```
+
+This change is guarded by a language version. Any class declaration in a library
+whose language version is before the one this feature ships in can be used as a
+mixin as long as the class meets the mixin restrictions. Classes in libraries
+whose version is later than that must be explicitly marked `mixin class` to
+allow the class to be used in a `with` clause.
+
+When upgrading your library to the new language version, you can preserve the
+existing behavior by adding `mixin` to every class declaration that can be used
+as a mixin. If the class defines a generative constructor or extends anything
+other than `Object`, then it already can't be used as a mixin and no change is
+needed.
+
+## Syntax
+
+This proposal builds on the existing sealed types proposal so the grammar
+includes those changes. The full set of modifiers that can appear before a class
+or mixin are `abstract`, `sealed`, `base`, `interface`, `final`, and `mixin`.
+Some combinations don't make sense:
+
+*   `sealed` implies `abstract`, so they can't be combined.
+*   `sealed` implies non-extensibility, so can't be combined with `interface`
+    or `final`.
+*   `sealed` implies non-implementability, so can't be combined with `base` or
+    `final`.
+*   `base`, `interface`, and `final` all control the same two capabilities so
+    are mutually exclusive.
+*   `mixin` as a modifier can obviously only be applied to a class. It can be
+    combined with any other modifiers that can be applied to a class.
+
+The remaining valid combinations are:
+
+<table><tr><td>
+
+```
+class
+sealed class
+base class
+interface class
+final class
+abstract class
+abstract base class
+abstract interface class
+abstract final class
+```
+
+</td><td>
+
+```
+mixin class
+sealed mixin class
+base mixin class
+interface mixin class
+final mixin class
+abstract mixin class
+abstract base mixin class
+abstract interface mixin class
+abstract final mixin class
+```
+
+</td><td>
+
+```
+mixin
+sealed mixin
+base mixin
+interface mixin
```

Comment:
We've been thinking of the two capabilities that `base`/`interface`/`final` controls as:

* Can be extended 
* Can be implemented

After thinking through the invariants we care about and the combinations users might want to express, I think it makes more sense to think of the two capabilities as:

* Can be used as a superclass
* Can be implemented

When a type is marked `final` or `interface`, it loses the ability to be used as a superclass. That means:

* For a `class`, it can't be extended.
* For a `mixin`, it can't be mixed in.

At first, I thought it was stupid to have a mixin declaration that you can't mixin, but it still be mixed in *inside* the library, so I think it's useful to be able to express "mixin that can be implemented but not mixed in" and "mixin that can't be implemented or mixed in". Treating the capability as "can be used as a superclass" lets us express that with the existing modifiers:

```
mixin // Can mix in and implement.
base mixin // Can mix in but not implement.
interface mixin // Can implement but not mix in.
final mixin // Can't implement or mix in.
```

Then there's the question of `mixin class`. For those, I think it simply doesn't make sense to combine a modifier that takes away subclassing or mixin in. `final mixin class` doesn't make sense because, as you note, you could just do `extends Object with` instead. And if you want to prevent both extending and using it as a mixin... then you shouldn't have put `mixin` on it.

Reply:
Here I'd say that the `final` applies to both the `mixin` declaration and the `class` declaration of the `mixin class`. It can neither be implemented or used as a superclass outside of the library.
It can be used as either inside the library.

That only makes sense if you cannot mix in a non-`mixin` `class` inside the same library, which I'm also proposing.